### PR TITLE
Wait for dblClick before caret moving to support numberBox selection on double click for ie11(T698588) 

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: js/ui/number_box/number_box.mask.js [hunk 1]
```diff
@@ -25,6 +25,8 @@ var ensureDefined = function(value, defaultValue) {
     return value === undefined ? defaultValue : value;
 };
 
+var caretTimeoutDuration = browser.msie ? 300 : 0;
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]: upper case

File: js/ui/number_box/number_box.mask.js [hunk 2]
```diff
@@ -72,12 +74,15 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
 
-        if(browser.msie) {
-            clearTimeout(this._ieCaretTimeout);
-            this._ieCaretTimeout = setTimeout(this._moveCaretToBoundary.bind(this, MOVE_BACKWARD, e));
-        } else {
-            this._moveCaretToBoundary(MOVE_BACKWARD, e);
+        var that = this;
+        if(this._caretTimeout) {
+            clearTimeout(this._caretTimeout);
+            this._caretTimeout = null;
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]: remove

File: js/ui/number_box/number_box.mask.js [hunk 3]
```diff
@@ -72,12 +74,15 @@ var NumberBoxMask = NumberBoxBase.inherit({
             return;
         }
 
-        if(browser.msie) {
-            clearTimeout(this._ieCaretTimeout);
-            this._ieCaretTimeout = setTimeout(this._moveCaretToBoundary.bind(this, MOVE_BACKWARD, e));
-        } else {
-            this._moveCaretToBoundary(MOVE_BACKWARD, e);
+        var that = this;
+        if(this._caretTimeout) {
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]: `this._caretTimeout && clearTimeout(this._caretTimeout);`

File: js/ui/number_box/number_box.mask.js [hunk 4]
```diff
@@ -72,12 +74,15 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
+            clearTimeout(this._caretTimeout);
+            this._caretTimeout = null;
         }
+        this._caretTimeout = setTimeout(function() {
+            that._caretTimeout = undefined;
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]: `null`

File: js/ui/number_box/number_box.mask.js [hunk 5]
```diff
@@ -72,12 +74,15 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
+            that._moveCaretToBoundary(MOVE_BACKWARD, e);
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]: remove `e`

File: js/ui/number_box/number_box.mask.js [hunk 6]
```diff
@@ -457,12 +467,33 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
+    },
+
     _attachFormatterEvents: function() {
         var $input = this._input();
 
         eventsEngine.on($input, eventUtils.addNamespace(INPUT_EVENT, NUMBER_FORMATTER_NAMESPACE), this._formatValue.bind(this));
         eventsEngine.on($input, eventUtils.addNamespace("dxclick", NUMBER_FORMATTER_NAMESPACE), function() {
-            this._caret(maskCaret.getCaretInBoundaries(this._caret(), this._getInputVal(), this._getFormatPattern()));
+            var that = this;
+
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]:
```javascript
var shouldReloadCaretTimer = !!this._caretTimeout;
```

File: js/ui/number_box/number_box.mask.js (hunk 1 above)
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]:
```javascript
var doesBrowserClearSelectionIfCaretMovedAfterClick = browser.msie;
var caretTimeoutDuration = doesBrowserClearSelectionIfCaretMovedAfterClick  ? 300 : 0;
```

File: js/ui/number_box/number_box.mask.js [hunk 7]
```diff
@@ -457,12 +467,33 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
+                    that._caretTimeout = null;
+                    that._moveCaretToRightPositionAfterFormatting();
+                }, caretTimeoutDuration);
+            } else {
+                that._moveCaretToRightPositionAfterFormatting();
+            }
+        }.bind(this));
+
+        eventsEngine.on($input, "dxdblclick", function() {
+            if(this._caretTimeout) {
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js]:
```javascript
this._clearCaretTimeout() {
   this._caretTimeout &&  clearTimeout(this._caretTimeout);
   this._caretTimeout = null;
}
```

File: js/ui/number_box/number_box.mask.js [hunk 8]
```diff
@@ -542,8 +573,8 @@ var NumberBoxMask = NumberBoxBase.inherit({
                 var caretInBoundaries = maskCaret.getCaretInBoundaries(caret, currentText, format);
 
                 if(browser.msie) {
```
Comment (churkin) [inline js/ui/number_box/number_box.mask.js:563]: extra condition

**Review by churkin — CHANGES_REQUESTED** [review]

File: js/ui/number_box/number_box.mask.js [hunk 9]
```diff
@@ -72,12 +74,13 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
-        } else {
-            this._moveCaretToBoundary(MOVE_BACKWARD, e);
-        }
+        var that = this;
+
+        this.clearCaretTimeout();
+        this._caretTimeout = setTimeout(function() {
+            that._caretTimeout = null;
+            that._moveCaretToBoundaryEventHandler(MOVE_BACKWARD, e);
+        }, CARET_TIMEOUT_DURATION);
```
Comment (DokaRus) [inline js/ui/number_box/number_box.mask.js]:
```suggestion
        function() {
          this._caretTimeout = null;
          this._moveCaretToBoundaryEventHandler(MOVE_BACKWARD, e);
        }.bind(this)
```

File: js/ui/number_box/number_box.mask.js [hunk 10]
```diff
@@ -462,8 +470,23 @@ var NumberBoxMask = NumberBoxBase.inherit({
 
         eventsEngine.on($input, eventUtils.addNamespace(INPUT_EVENT, NUMBER_FORMATTER_NAMESPACE), this._formatValue.bind(this));
         eventsEngine.on($input, eventUtils.addNamespace("dxclick", NUMBER_FORMATTER_NAMESPACE), function() {
-            this._caret(maskCaret.getCaretInBoundaries(this._caret(), this._getInputVal(), this._getFormatPattern()));
+            var that = this;
+
+            if(!this._caretTimeout) {
+                this._caretTimeout = setTimeout(function() {
+                    that._caret(maskCaret.getCaretInBoundaries(that._caret(), that._getInputVal(), that._getFormatPattern()));
+                }, CARET_TIMEOUT_DURATION);
```
Comment (DokaRus) [inline js/ui/number_box/number_box.mask.js]: function() {).bind(this)
```suggestion
                function() {
                  this._caret(maskCaret.getCaretInBoundaries(this._caret(), this._getInputVal(), this._getFormatPattern()));
                ).bind(this)
```

File: js/ui/number_box/number_box.mask.js [hunk 11]
```diff
@@ -72,12 +74,13 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
-        if(browser.msie) {
-            clearTimeout(this._ieCaretTimeout);
-            this._ieCaretTimeout = setTimeout(this._moveCaretToBoundary.bind(this, MOVE_BACKWARD, e));
-        } else {
-            this._moveCaretToBoundary(MOVE_BACKWARD, e);
-        }
+        var that = this;
+
+        this.clearCaretTimeout();
+        this._caretTimeout = setTimeout(function() {
```
Comment (DokaRus) [inline js/ui/number_box/number_box.mask.js:78]: Note about tests: you should update clock after using KeyboardMock methods like `type`, `press` etc.

**Review by churkin — APPROVED** [review]

**Review by DokaRus — APPROVED** [review]

**Pull request merged**